### DISCUSSION TITLE
motion_detected fails with exception - "RuntimeError: deque mutated during iteration"

### DISCUSSION THREAD

**Discussion by dbrandesky** [gpiozero/gpiozero#973] [Q&A]
I have a fairly "simple" loop on a Raspberry Pi Zero W that runs two applications: the first basically generates random "arrivals" of a simulated train and activates GPIO pins when those events occur; the second is a function that monitors the state of an external motion detector, and passes that state into a function which will return True if motion is detected.

The exception is as follows:

```python
Traceback (most recent call last):
  File "TrainGenerator.py", line 107, in <module>
    detector_state = garage_motion(mqttc, topic_state, motion_sensor
)
  File "TrainGenerator.py", line 54, in garage_motion
    if sensor.motion_detected:
  File "/home/pi/TrainGenerator/venv/lib/python3.7/site-packages/gpi
ozero/input_devices.py", line 350, in is_active
    return self.value > self.threshold   
  File "/home/pi/TrainGenerator/venv/lib/python3.7/site-packages/gpi
ozero/input_devices.py", line 604, in value
    return super(MotionSensor, self).value
  File "/home/pi/TrainGenerator/venv/lib/python3.7/site-packages/gpi
ozero/input_devices.py", line 325, in value
    return self._queue.value
  File "/home/pi/TrainGenerator/venv/lib/python3.7/site-packages/gpi
ozero/mixins.py", line 587, in value
    return self.average(self.queue)
  File "/usr/lib/python3.7/statistics.py", line 311, in mean
    T, total, count = _sum(data)
  File "/usr/lib/python3.7/statistics.py", line 147, in _sum
    for n,d in map(_exact_ratio, values):
RuntimeError: deque mutated during iteration
```

The offending function is this:

```python
def garage_motion(mqtt_client, topic, sensor):
    if sensor.motion_detected:
        mqtt_publish(mqtt_client, topic, 'ON')
        return True
    else:
        mqtt_publish(mqtt_client, topic, 'OFF')
        return False
```

This function is run every 0.1 seconds in the main loop.

Full code can be found [here](https://pastebin.com/6zXvvWnX).

I'm totally at a loss here, any tips would be appreciated.

EDIT: I have no idea why the line breaks all got lost in my code snippets.

**Comment by bennuttall** [comment]
> EDIT: I have no idea why the line breaks all got lost in my code snippets.

Use three backticks to fence your code, not one (I've edited it)

**Comment by bennuttall** [comment]
What version of gpiozero is this?

**Comment by dbrandesky** [comment]
> What version of gpiozero is this?

Version 1.6.2 running on Python 3.7.3.

**Comment by bennuttall** [comment]
It looks like you're polling too fast for it to handle. You could try waiting a little longer.

Alternatively you could try another method of triggering changes from the sensor:

```python
sensor.when_motion = publish_on
sensor.when_no_motion = publish_off
```

or

```python
while True:
    sensor.wait_for_motion()
    mqtt_publish(mqtt_client, topic, 'ON')
    sensor.wait_for_no_motion()
    mqtt_publish(mqtt_client, topic, 'OFF')
```

**Comment by dbrandesky** [comment]
> It looks like you're polling too fast for it to handle. You could try waiting a little longer.
> 
> Alternatively you could try another method of triggering changes from the sensor:
> 
> ```python
> sensor.when_motion = publish_on
> sensor.when_no_motion = publish_off
> ```
> 
> or
> 
> ```python
> while True:
>     sensor.wait_for_motion()
>     mqtt_publish(mqtt_client, topic, 'ON')
>     sensor.wait_for_no_motion()
>     mqtt_publish(mqtt_client, topic, 'OFF')
> ```

I definitely don't mind a longer wait. I currently wait for 0.1 seconds, and the code ran for about two days before the exception in that test.

I wasn't really clear on the wait_for_motion() method; it seemed like it was a blocking call; i.e. my loop would pause until motion occurred.

**Comment by bennuttall** [comment]
> I wasn't really clear on the wait_for_motion() method; it seemed like it was a blocking call; i.e. my loop would pause until motion occurred.

Yes, it's quite useful if that's compatible with your program flow. There are 4 different ways of reacting to device changes: polling, blocking, event-driven and source/values. See the [FAQ](https://gpiozero.readthedocs.io/en/stable/faq.html#what-s-the-difference-between-when-pressed-is-pressed-and-wait-for-press) which covers this topic for buttons, and it's much the same for other devices.

**Comment by lurch** [comment]
Looking at the line that caused the exception:
```
self.average(self.queue)
```
...I'm not sure if any of @bennuttall 's workarounds would actually "fix" the problem?
Because a MotionSensor has "noisy" / "jittery" output, GpioZero runs a background thread collecting (noisy) values into `self.queue`, and when you try to get the `.value` it returns the (smooth) average of the values in the queue. So the bug is that if you're _very_ unlucky, the background thread might be attempting to add a new value to the queue at _exactly_ the same time as the average is being calculated, which is what then leads to the `deque mutated during iteration` error. (and the chances of this bug occurring will be greater on a PiZero than on a Pi4, because the slower CPU in the PiZero will take (slightly) longer to run the `average` function)

I think the fix is therefore for https://github.com/gpiozero/gpiozero/blob/master/gpiozero/mixins.py#L538 to implement some kind of locking mechanism, so that the `fill` function doesn't try to append a value to the queue while the `value` function is calculating the average of the queue.
His code is normally so comprehensive, that there's part of me that's surprised that @waveform80 didn't already add this locking :wink:

**Reply by waveform80** [reply]
@lurch is absolutely right -- the fact this is a deque mutation error from within gpiozero's code pretty much screams this is a bug. However, in this case I know why this has suddenly appeared and it's horribly subtle (and yes, a case of me cutting corners!)...

Prior to 1.6.x, the default average routine for `MotionSensor` (or any derivative of `SmoothedInputDevice`) was median. Now this never caused an issue because the first thing in the process of calculating the median is to sort the list. The process of sorting the list first duplicates it. Hang on! Doesn't that involve iteration? Well, yes, but within an explicitly locked portion of the interpreter so it'll never get interrupted by another Python thread. But in 1.6.x the default average for `MotionSensor` changed to `mean` (can't seem to find the thread/ticket where that got hashed out, but it had to do with giving `MotionSensor` a less "binary" value as I recall). Anyway, `mean` (obviously) doesn't need to sort and so merrily iterates over the deque (which may be mutated during said iteration).

The fix is (as @lurch suggests), throw a lock around the call to `self.average` (or pass an atomic (as far as Python is concerned) copy of the deque to `self.average`, although frankly I lean towards the lock as a lighter-weight alternative).

Anyway, I'll re-open the original ticket and assign it to me in a mo.

**Reply by waveform80** [reply]
Oh ... apparently I can't re-open the issue. Oh well, @bennuttall any chance you can dupe this into another issue and assign it to me so I don't forget it?
